asserts: stm32: Adding asserts enable/disable

ikmdani opened the pull request:
This commit allows to enable asserts functionality for stm32 series.
STM32Cube hal & ll drivers provides asserts and the same can be 
enabled using a Kconfig symbol USE_STM32_ASSERT.

Depends on https://github.com/zephyrproject-rtos/hal_stm32/pull/100

Signed-off-by: Krishna Mohan Dani <krishnamohan.d@hcl.com>

ikmdani committed:
asserts: stm32: Adding asserts

This commit adds the asserts symbol in Kconfig to enable/disable
asserts functionality for stm32 series. These would be used in
stm32cube hal & ll drivers.

Signed-off-by: Krishna Mohan Dani <krishnamohan.d@hcl.com>

FRASTM reviewed (approved):
Should the `USE_STM32_ASSERT` definition also imply setting the **USE_FULL_ASSERT** in the modules/hal/stm32/stm32cube/stm32XXxx/drivers/include/stm32XXxx_hal_conf.h

An example could help demonstrating the assertion.

ABOSTM reviewed (approved):
LGTM

erwango reviewed (changes requested):
This miss a link to matching hal_stm32 PR

ikmdani commented:
> This miss a link to matching hal_stm32 PR

Updated the PR description with a mention of dependency.

erwango commented:
> > This miss a link to matching hal_stm32 PR
> 
> Updated the PR description with a mention of dependency.

Sorry, I missed to povide the link: 
https://docs.zephyrproject.org/latest/guides/modules.html#submitting-changes-to-modules

ikmdani committed:
west: update west.yml to add updated hal_stm32 modules

This commit changes the entry referencing the hal_stm32 module
into west.yml. This brings in the asserts feature from stm32cube
hal/ll and zephyr.

Signed-off-by: Krishna Mohan Dani <krishnamohan.d@hcl.com>

ikmdani commented:
> This miss a link to matching hal_stm32 PR

https://github.com/zephyrproject-rtos/zephyr/pull/36469/commits/cc7b398fec18baab0ae85b97002c86b690fe5a1e

erwango commented on soc/arm/st_stm32/common/stm32cube_hal.c:
```diff
@@ -35,3 +35,18 @@ void HAL_Delay(__IO uint32_t Delay)
+
+#ifdef CONFIG_USE_STM32_ASSERT
+/**
+ * @brief This function calls zephyr's assert function if the caller's
+ * expression failed in assert_param() function.
```
```suggestion
 * @brief Generates an assert on STM32Cube HAL/LL assert trigger.
```

erwango commented on soc/arm/st_stm32/common/stm32cube_hal.c:
```diff
@@ -35,3 +35,18 @@ void HAL_Delay(__IO uint32_t Delay)
+ */
+void assert_failed(uint8_t *file, uint32_t line)
+{
+	/* expression is being set to 0 */
```
```suggestion
	/* Assert condition have been verified at Cube level, force generation here. */
```

erwango commented on soc/arm/st_stm32/common/stm32cube_hal.c:
```diff
@@ -35,3 +35,18 @@ void HAL_Delay(__IO uint32_t Delay)
+void assert_failed(uint8_t *file, uint32_t line)
+{
+	/* expression is being set to 0 */
+	__ASSERT(0, "Invalid value line %d @ %s\n", line, file);
```
```suggestion
	__ASSERT(false, "Invalid value line %d @ %s\n", line, file);
```
